Automattic/wp-calypso
# Plans 2023: Add pricing to Plans data-store and use in plans-grid

Related to (and may potentially close) https://github.com/Automattic/wp-calypso/issues/79005

## Proposed Changes

These changes are about adding pricing data to the Plans data-store and using it in plans-grid component (package). This is one of the bigger/most-critical steps in the NPM migration of the pricing tables (PT: pau2Xa-59h-p2 ). From this point on, plans-grid would have no dependency on Calypso state selectors for pricing (and some parts of the existing state ought to be marked as deprecated to promote the use of Plans data-store for everything else)

### Notes about implementation

- We transform the API data to an expected structure for pricing at the root, so that's the only structure we see when debugging or developing. This may have its drawbacks, although some level of abstraction is probably for the better.
- The pricing is doubly-projected for the plans-grid through `usePricingMetaForGridPlans`, which gives a finer representation of pricing (over the same data-store structure) based on how it's expected for the UI.
- The root-level representation takes a simple form: `originalPrice`, `discountedPrice`. This is what's been proven to work for all the needs of the pricing tables and should be applicable everywhere, even without additional selectors. A distinction of "discounts" comes from the source of the pricing meta:
    - If it's read over `usePlans`, then the discount is a technical "currency conversion" based one
    - If it's read over `useSitePlans`, then the discount is a prorated price
    - `originalPrice` should be the same for both

### Best be done in follow-ups (create issues)

1. Address remaining dependencies and migrate `usePricingMetaForGridPlans` to packages (either the existing plans-grid package or into data-stores - probably the former). This ought to happen with some **priority**.
2. Additional refactors ought to happen in the data-store
    * Clean the clutter from previously: remove the wp-data store, update types to use the new interfaces being redone/introduced here, etc.
3. Mark as deprecated the existing Calypso state selectors and/or replace with the data store's
    * `getPlanPrices` selector should be the first to go, with respective tests migrated to the data-store. Some **priority** here too.
4. Add unit tests and potentially create more fine-grained selectors for pricing (if we absolutely need any)

## Testing Instructions



1. Code review - if things/structures make sense
2. Unit tests - some tests were updated
3. Manual testing. Need to test all the cases of discounted prices across Admin & Signup contexts:

- In `/start/plans` and any tailored flow:
    - confirm that regular prices show without discounts for USD/EUR currencies
    - confirm currency discount is shown for INR currency (set through SA)
- In `/plans/[free]`
   - Confirm the same as above
- In `/plans/[paid]`
   - Confirm "upgrade credits" badge shows and discounted pricing is correct
- For intro offers:
   - Go through instructions in https://github.com/Automattic/wp-calypso/pull/84763 and confirm intro offers are shown correctly (or compared with production)
- Ensure large/small font displays for the prices render correctly as in production (for both intro offers and regular pricing)

## Pre-merge Checklist
